Installer/Read/manpages/novacli.1
.TH NOVACLI "1" "February 2012" "12.02"

.SH NAME
.B novacli
\- The Nova command line interface

.SH SYNOPSIS
.B novacli

.SH DESCRIPTION
.PP
.I novacli
is the command line interface for controlling and communicating with the Nova daemon. The CLI has the ability to start and stop the haystack and the Nova daemon, and also has the ability to query it for status and state information. Run novacli with the --help option to display usage options.

.TP
status nova|haystack
Outputs if the nova or haystack process is running and responding

.TP
novacli start nova|haystack
Starts the nova or haystack process

.TP
novacli stop nova|haystack
Stops the nova or haystack process

.TP
novacli list all|hostile|benign
Outputs the current list of suspect IP addresses

.TP
novacli get all|hostile|benign [csv]
Outputs the details for all suspects of a type (all, hostile only, or benign only). Optionally can be output in CSV format.

.TP
novacli get xxx.xxx.xxx.xxx
Outputs the details of a specific suspect with IP address xxx.xxx.xxx.xxx

.TP
novacli clear all
Clears all saved data for suspects

.TP
novacli clear xxx.xxx.xxx.xxx
Clears all saved data for a specific suspect


.SH COPYRIGHT
.PP
Nova is Copyright (c) 2011 - 2012 DataSoft Corporation
.PP
Nova is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. Nova is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with Nova. If not, see <http://www.gnu.org/licenses/>.


debian/changelog
nova (12.6-1) unstable; urgency=low

  * Beta Release: 12.6, May 2012 release.
  * Many performance improvements
  * Upgraded messaging subsystem
  * More featured WebUI

 -- Dan Petro <dan.petro@datasoft.com>  Fri, 01 Jun 2012 10:17:16 -0700


NovaCLI/src/NovaCLI.cpp
//============================================================================
// Name        : NovaCLI.cpp
// Copyright   : DataSoft Corporation 2011-2012
//	Nova is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.
//
//   Nova is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with Nova.  If not, see <http://www.gnu.org/licenses/>.
// Description : Command line interface for Nova
//============================================================================

#include "NovaCLI.h"
#include "nova_ui_core.h"
#include "Logger.h"


#include <iostream>
#include <stdlib.h>
#include "inttypes.h"
#include "boost/program_options.hpp"


namespace po = boost::program_options;
using namespace std;
using namespace Nova;
using namespace NovaCLI;


int main(int argc, const char *argv[])
{
	// Fail if no arguements
	if(argc < 2)
	{
		PrintUsage();
	}

	Config::Inst();

	// Disable notifications and email in the CLI
	Logger::Inst()->SetUserLogPreferences(LIBNOTIFY, EMERGENCY, '+');
	Logger::Inst()->SetUserLogPreferences(EMAIL, EMERGENCY, '+');
	InitializeUI();

	// We parse the input arguments here,
	// but refer to other functions to do any
	// actual work.

	// Listing suspect IP addresses
	if(!strcmp(argv[1], "list"))
	{
		if(argc < 3)
		{
			PrintUsage();
		}

		if(!strcmp(argv[2], "all"))
		{
			PrintSuspectList(SUSPECTLIST_ALL);
		}
		else if(!strcmp(argv[2], "hostile"))
		{
			PrintSuspectList(SUSPECTLIST_HOSTILE);
		}
		else if(!strcmp(argv[2], "benign"))
		{
			PrintSuspectList(SUSPECTLIST_BENIGN);
		}
		else
		{
			PrintUsage();
		}
	}

	// Checking status of components
	else if(!strcmp(argv[1], "status"))
	{
		if(argc < 3)
		{
			PrintUsage();
		}

		if(!strcmp(argv[2], "nova"))
		{
			StatusNovaWrapper();
		}
		else if(!strcmp(argv[2], "haystack"))
		{
			StatusHaystackWrapper();
		}
		else
		{
			PrintUsage();
		}
	}

	// Starting components
	else if(!strcmp(argv[1], "start"))
	{
		if(argc < 3)
		{
			PrintUsage();
		}

		if(!strcmp(argv[2], "nova"))
		{
			StartNovaWrapper();
		}
		else if(!strcmp(argv[2], "haystack"))
		{
			StartHaystackWrapper();
		}
		else
		{
			PrintUsage();
		}
	}

	// Stopping components
	else if(!strcmp(argv[1], "stop"))
	{
		if(argc < 3)
		{
			PrintUsage();
		}

		if(!strcmp(argv[2], "nova"))
		{
			StopNovaWrapper();
		}
		else if(!strcmp(argv[2], "haystack"))
		{
			StopHaystackWrapper();
		}
		else
		{
			PrintUsage();
		}
	}

	// Getting suspect information
	else if(!strcmp(argv[1], "get"))
	{
		if(argc < 3)
		{
			PrintUsage();
		}

		if(!strcmp(argv[2], "all"))
		{
			if(argc == 4 && !strcmp(argv[3], "csv"))
			{
				PrintAllSuspects(SUSPECTLIST_ALL, true);
			}
			else
			{
				PrintAllSuspects(SUSPECTLIST_ALL, false);
			}
		}
		else if(!strcmp(argv[2], "hostile"))
		{
			if(argc == 4 && !strcmp(argv[3], "csv"))
			{
				PrintAllSuspects(SUSPECTLIST_HOSTILE, true);
			}
			else
			{
				PrintAllSuspects(SUSPECTLIST_HOSTILE, false);
			}
		}
		else if(!strcmp(argv[2], "benign"))
		{
			if(argc == 4 && !strcmp(argv[3], "csv"))
			{
				PrintAllSuspects(SUSPECTLIST_BENIGN, true);
			}
			else
			{
				PrintAllSuspects(SUSPECTLIST_BENIGN, false);
			}
		}
		else
		{
			// Some early error checking for the
			in_addr_t address;
			if(inet_pton(AF_INET, argv[2], &address) != 1)
			{
				cout << "Error: Unable to convert to IP address" << endl;
				exit(EXIT_FAILURE);
			}

			PrintSuspect(address);
		}
	}

	// Clearing suspect information
	else if(!strcmp(argv[1], "clear"))
	{
		if(argc < 3)
		{
			PrintUsage();
		}

		if(!strcmp(argv[2], "all"))
		{
			ClearAllSuspectsWrapper();
		}
		else
		{
			// Some early error checking for the
			in_addr_t address;
			if(inet_pton(AF_INET, argv[2], &address) != 1)
			{
				cout << "Error: Unable to convert to IP address" << endl;
				exit(EXIT_FAILURE);
			}

			ClearSuspectWrapper(address);
		}
	}

	// Checking status of components
	else if(!strcmp(argv[1], "uptime"))
	{
		PrintUptime();
	}


	else if(!strcmp(argv[1], "readsetting"))
	{
		if(argc < 3)
		{
			PrintUsage();
		}

		cout << Config::Inst()->ReadSetting(string(argv[2])) << endl;
	}

	else if(!strcmp(argv[1], "writesetting"))
	{
		if(argc < 4)
		{
			PrintUsage();
		}

		cout << Config::Inst()->WriteSetting(string(argv[2]), string(argv[3])) << endl;
	}

	else if(!strcmp(argv[1], "listsettings"))
	{
		vector<string> settings = Config::Inst()->GetPrefixes();
		for (uint i = 0; i < settings.size(); i++)
		{
			cout << settings[i] << endl;
		}
	}

	else
	{
		PrintUsage();
	}

	return EXIT_SUCCESS;
}

namespace NovaCLI
{

void PrintUsage()
{
	cout << "Usage:" << endl;
	cout << "    " << EXECUTABLE_NAME << " status nova|haystack" << endl;
	cout << "    " << EXECUTABLE_NAME << " start nova|haystack" << endl;
	cout << "    " << EXECUTABLE_NAME << " stop nova|haystack" << endl;
	cout << "    " << EXECUTABLE_NAME << " list all|hostile|benign" << endl;
	cout << "    " << EXECUTABLE_NAME << " get all|hostile|benign [csv]" << endl;
	cout << "    " << EXECUTABLE_NAME << " get xxx.xxx.xxx.xxx" << endl;
	cout << "    " << EXECUTABLE_NAME << " clear all" << endl;
	cout << "    " << EXECUTABLE_NAME << " clear xxx.xxx.xxx.xxx" << endl;
	cout << "    " << EXECUTABLE_NAME << " writesetting SETTING VALUE" << endl;
	cout << "    " << EXECUTABLE_NAME << " readsetting SETTING" << endl;
	cout << "    " << EXECUTABLE_NAME << " listsettings" << endl;
	cout << endl;

	exit(EXIT_FAILURE);
}

void StatusNovaWrapper()
{
	if(!ConnectToNovad())
	{
		cout << "Novad Status: Not running" << endl;
	}
	else
	{
		if(IsNovadUp(false))
		{
			cout << "Novad Status: Running and responding" << endl;
		}
		else
		{
			cout << "Novad Status: Not responding" << endl;
		}

		CloseNovadConnection();
	}
}

void StatusHaystackWrapper()
{
	if(IsHaystackUp())
	{
		cout << "Haystack Status: Running" << endl;
	}
	else
	{
		cout << "Haystack Status: Not running" << endl;
	}
}

void StartNovaWrapper()
{
	if(!ConnectToNovad())
	{
		if(StartNovad())
		{
			cout << "Started Novad" << endl;
		}
		else
		{
			cout << "Failed to start Novad" << endl;
		}
	}
	else
	{
		cout << "Novad is already running" << endl;
		CloseNovadConnection();
	}
}

void StartHaystackWrapper()
{
	if(!IsHaystackUp())
	{
		if(StartHaystack())
		{
			cout << "Started Haystack" << endl;
		}
		else
		{
			cout << "Failed to start Haystack" << endl;
		}
	}
	else
	{
		cout << "Haystack is already running" << endl;
	}
}

void StopNovaWrapper()
{
	Connect();

	if(StopNovad())
	{
		cout << "Novad has been stopped" << endl;
	}
	else
	{
		cout << "There was a problem stopping Novad" << endl;
	}
}

void StopHaystackWrapper()
{
	if(StopHaystack())
	{
		cout << "Haystack has been stopped" << endl;
	}
	else
	{
		cout << "There was a problem stopping the Haystack" << endl;
	}
}

void PrintSuspect(in_addr_t address)
{
	Connect();

	Suspect *suspect = GetSuspect(ntohl(address));

	if(suspect != NULL)
	{
		cout << suspect->ToString() << endl;
	}
	else
	{
		cout << "Error: No suspect received" << endl;
	}

	delete suspect;

	CloseNovadConnection();
}

void PrintAllSuspects(enum SuspectListType listType, bool csv)
{
	Connect();

	vector<in_addr_t> *suspects;
	suspects = GetSuspectList(listType);

	if(suspects == NULL)
	{
		cout << "Failed to get suspect list" << endl;
		exit(EXIT_FAILURE);
	}

	for(uint i = 0; i < suspects->size(); i++)
	{
		Suspect *suspect = GetSuspect(suspects->at(i));

		if(suspect != NULL)
		{
			if(!csv)
			{
				cout << suspect->ToString() << endl;
			}
			else
			{
				cout << suspect->GetIpString() << ",";
				for(int i = 0; i < DIM; i++)
				{
					cout << suspect->GetFeatureSet().m_features[i] << ",";
				}
				cout << suspect->GetClassification() << endl;
			}

			delete suspect;
		}
		else
		{
			cout << "Error: No suspect received" << endl;
		}
	}

	CloseNovadConnection();

}

void PrintSuspectList(enum SuspectListType listType)
{
	Connect();

	vector<in_addr_t> *suspects;
	suspects = GetSuspectList(listType);


	if(suspects == NULL)
	{
		cout << "Failed to get suspect list" << endl;
		exit(EXIT_FAILURE);
	}

	for(uint i = 0; i < suspects->size(); i++)
	{
		in_addr tmp;
		tmp.s_addr = suspects->at(i);
		char *address = inet_ntoa(tmp);
		cout << address << endl;
	}

	CloseNovadConnection();
}

void ClearAllSuspectsWrapper()
{
	Connect();

	if(ClearAllSuspects())
	{
		cout << "Suspects have been cleared" << endl;
	}
	else
	{
		cout << "There was an error when clearing the suspects" << endl;
	}

	CloseNovadConnection();
}

void ClearSuspectWrapper(in_addr_t address)
{
	Connect();

	if(ClearSuspect(address))
	{
		cout << "Suspect data has been cleared for this suspect" << endl;
	}
	else
	{
		cout << "There was an errer when trying to clear the suspect data for this suspect" << endl;
	}

	CloseNovadConnection();
}

void PrintUptime()
{
	Connect();
	cout << "Uptime is: " << GetStartTime() << endl;
	CloseNovadConnection();
}

void Connect()
{
	if(!ConnectToNovad())
	{
		cout << "ERROR: Unable to connect to Nova" << endl;
		exit(EXIT_FAILURE);
	}
}

}
